### PR TITLE
feat(cbork): Add new cbor kit crates

### LINKED ISSUE #178: 🛠️ [TASK] : CBOR/CDDL/ABNF Parser crates

### PR DESCRIPTION
**Pull request by stevenj**
# Description

CBOR CDDL and ABNF Parsers

## Related Issue(s)

Closes #178

## Description of Changes

Initial CDDL/ABNF parser libraries for future CDDL tooling requirements

## Breaking Changes

None, these crates are currently unused.

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

File: hermes/crates/cbork/.config/nextest.toml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# cspell: words scrollability testcase
```
Comment (minikin) [inline hermes/crates/cbork/.config/nextest.toml:1]: Can we move cspell words here and in the other files to [project.dic](https://github.com/input-output-hk/hermes/blob/main/.config/dictionaries/project.dic)?

**Review by minikin — CHANGES_REQUESTED** [review]
PRs of this size are tricky to review. We need to find a way to make PRs smaller in the future.

**Review by saibatizoku — DISMISSED** [review]

File: hermes/crates/cbork/.config/nextest.toml (hunk 1 above)
Comment (stevenj) [reply hermes/crates/cbork/.config/nextest.toml:1]: `testcase` is general enough it should be in `project.dic`.
We should only really use local word lists for words that are unusual and self contained in the file.
Author surnames is a common example I come across.
`scrollability` can just be changed in the source to `scroll ability` even if it came from some other source.
Check the other spellings for these kinds of things.

File: hermes/crates/cbork/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/Cargo.toml:1]:
```suggestion
```

Add `abnf` to project words

File: hermes/crates/cbork/Earthfile [hunk 3]
```diff
@@ -0,0 +1,44 @@
+VERSION 0.7
+
+#cspell: words rustfmt abnf
```
Comment (stevenj) [inline hermes/crates/cbork/Earthfile:3]:
```suggestion
```

Add `rustfmt` to project words

File: hermes/crates/cbork/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<!-- cspell: words cbork CBOR -->
```
Comment (stevenj) [inline hermes/crates/cbork/README.md:1]:
```suggestion
```

Add `cbork` and `CBOR` to project words

File: hermes/crates/cbork/abnf-parser/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#cspell: words memchr abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/Cargo.toml:1]:
```suggestion
#cspell: words memchr
```

File: hermes/crates/cbork/abnf-parser/src/grammar/rfc_5234.pest [hunk 6]
```diff
@@ -0,0 +1,89 @@
+//! ABNF Grammar adapted from RFC5234 Section 4 and Appendix B
+//! https://datatracker.ietf.org/doc/html/rfc5234#section-4
+//! https://datatracker.ietf.org/doc/html/rfc5234#appendix-B.1
+
+// cspell: words ABNF abnf rulename VCHAR DQUOTE ONEOF
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/src/grammar/rfc_5234.pest:5]:
```suggestion
// cspell: words rulename VCHAR DQUOTE ONEOF
```

File: hermes/crates/cbork/abnf-parser/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,95 @@
+// cspell: words apskhem ABNF abnf Naur
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/src/lib.rs:1]:
```suggestion
// cspell: words Naur
```
add yourself to the project words

File: hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf [hunk 8]
```diff
@@ -0,0 +1,16 @@
+; cspell: words fullyear mday secfrac numoffset
+
+date-fullyear   = 4DIGIT
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf:3]:
```suggestion
date-full-year   = 4DIGIT
```

File: hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+date-month      = 2DIGIT  ; 01-12
+date-mday       = 2DIGIT  ; 01-28, 01-29, 01-30, 01-31 based on month/year
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf:5]:
```suggestion
date-m-day       = 2DIGIT  ; 01-28, 01-29, 01-30, 01-31 based on month/year
```

File: hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+time-hour       = 2DIGIT  ; 00-23
+time-minute     = 2DIGIT  ; 00-59
+time-second     = 2DIGIT  ; 00-58, 00-59, 00-60 based on leap sec rules
+time-secfrac    = "." 1*DIGIT
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf:9]:
```suggestion
time-sub-second    = "." 1*DIGIT
```

File: hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+time-numoffset  = ("+" / "-") time-hour ":" time-minute
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf:10]:
```suggestion
time-numeric-offset  = ("+" / "-") time-hour ":" time-minute
```

File: hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf [hunk 12]
```diff
@@ -0,0 +1,16 @@
+; cspell: words fullyear mday secfrac numoffset
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc3339.abnf:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc5234.abnf [hunk 13]
```diff
@@ -0,0 +1,61 @@
+; cspell: words rulelist rulename VCHAR DQUOTE VCHAR ONEOF HEXDIG
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/abnf/valid_abnf_rfc5234.abnf:1]:
```suggestion
; cspell: words VCHAR DQUOTE VCHAR ONEOF HEXDIG
```

add rulelist and rulename to project words

File: hermes/crates/cbork/abnf-parser/tests/testsuite/abnf.rs [hunk 14]
```diff
@@ -0,0 +1,54 @@
+// cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/abnf.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/alternations.rs [hunk 15]
```diff
@@ -0,0 +1,42 @@
+// cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/alternations.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/character_sets.rs [hunk 16]
```diff
@@ -0,0 +1,30 @@
+// cspell: words abnf VCHAR vchar
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/character_sets.rs:1]:
```suggestion
// cspell: words VCHAR vchar
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/comments.rs [hunk 17]
```diff
@@ -0,0 +1,63 @@
+// cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/comments.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/common/mod.rs [hunk 18]
```diff
@@ -0,0 +1,19 @@
+// cspell: words abnf ABNF
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/common/mod.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/groups.rs [hunk 19]
```diff
@@ -0,0 +1,44 @@
+// cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/groups.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/identifiers.rs [hunk 20]
```diff
@@ -0,0 +1,24 @@
+// cspell: words abnf RULENAME rulename
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/identifiers.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/main.rs [hunk 21]
```diff
@@ -0,0 +1,13 @@
+// cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/main.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/repetitions.rs [hunk 22]
```diff
@@ -0,0 +1,44 @@
+// cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/repetitions.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/rules.rs [hunk 23]
```diff
@@ -0,0 +1,52 @@
+// cspell: words abnf rulename rulelist
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/rules.rs:1]:
```suggestion
```

File: hermes/crates/cbork/abnf-parser/tests/testsuite/values.rs [hunk 24]
```diff
@@ -0,0 +1,151 @@
+// cspell: words abnf
```
Comment (stevenj) [inline hermes/crates/cbork/abnf-parser/tests/testsuite/values.rs:1]:
```suggestion
```

File: hermes/crates/cbork/cddl-parser/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,165 @@
+// cspell: words apskhem
```
Comment (stevenj) [inline hermes/crates/cbork/cddl-parser/src/lib.rs:1]:
```suggestion
```

**Review by stevenj — COMMENTED** [review]
See requested changes, and add these words to project words and remove from all cspell comments

File: hermes/crates/cbork/.config/nextest.toml (hunk 1 above)
Comment (stevenj) [reply hermes/crates/cbork/.config/nextest.toml]: common spelling moved to project words

**Pull request merged**